pkg/controller/service/service_controller.go

```diff
@@ -164,6 +164,21 @@ func New(
 	s.serviceLister = serviceInformer.Lister()
 	s.serviceListerSynced = serviceInformer.Informer().HasSynced
 
+	nodeInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				s.nodeSyncLoop()
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				s.nodeSyncLoop()
```

Comment:
Does a heartbeat incur a new resource version? If not, could we compare the versions between `old` and `cur` and skip calling `nodeSyncLoop` if they match?

Reply:
>Is it possible to filter out the useless events by looking at specific fields? like unschedulable/ node taints / node status

That's what `nodeSyncLoop` [is doing today](https://github.com/kubernetes/kubernetes/blob/c29e1c62c5a84d66abdfed2a20c5607a120b8ebd/pkg/controller/service/service_controller.go#L666-L674): it looks at nodes only that meet [our predicate](https://github.com/kubernetes/kubernetes/blob/c29e1c62c5a84d66abdfed2a20c5607a120b8ebd/pkg/controller/service/service_controller.go#L627-L661) (schedulable, ready, and others). Additionally, if the current host set and the one looked at previously happen to be equal, we only try to update hosts for those services that we failed to update in the last iteration.

Writing this down, I realize that we have another issue with the PR at this stage: based on the numbers that @MrHohn provided above, large clusters would run `s.updateLoadBalancerHosts` on very node status update even if nothing else changed in the node set. That's essentially a no-op in the case where there are no services that failed to update; if we had any, however, we'd [invoke `UpdateLoadBalancer`](https://github.com/kubernetes/kubernetes/blob/c29e1c62c5a84d66abdfed2a20c5607a120b8ebd/pkg/controller/service/service_controller.go#L721) on every node update. Essentially, we'd reach out to the cloud API with high frequency at a point where that API possibly experienced issues.

So the least we'd have to do is limit invoking `updateLoadBalancerHosts` to the period node sync (i.e., once per `nodeSyncPeriod` of 100 seconds as of now). We could do that by introducing a flag that's set when `nodeSyncLoop` [is called via the `nodeSyncPeriod`](https://github.com/kubernetes/kubernetes/blob/c29e1c62c5a84d66abdfed2a20c5607a120b8ebd/pkg/controller/service/service_controller.go#L223) and disabled when our event handler triggered it.

Assuming for a moment we did that change, we'd still call `s.nodeLister.ListWithPredicate(getNodeConditionPredicate()` and `nodeSlicesEqualForLB` on every node (status) update while holding the lock on the hosts list. Not exactly sure where that puts us in terms of performance for larger clusters, but it seems like something we should be aware of.

Reply:
>To move forward without having too much change to the event handling / sync loop, what about simply reducing `nodeSyncPeriod` to something that we think is acceptable?

That was basically [my initial suggestion](https://github.com/kubernetes/cloud-provider/issues/38#issuecomment-510203144), though I additionally proposed to make `nodeSyncPeriod` configurable through a flag.

Reply:
My understanding is that we have a reasonable default today, and the flag would allow cluster operators to reduce the node sync period if it makes sense in their environment. Admittedly, that won't be helpful in very large clusters.

I agree that coming up with a manageable solution at the controller level seems preferable.

Reply:
Now that node heart beats are handled by lease API this change seems more reasonable. Thoughts?